test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,962 @@
+    uint256 public constant CYCLE_TARGET = 1300 ether; // 0.0001 ether, 1e14 wei
+    uint256 public constant MIN_PLEDGE_AMOUNT = 45 ether; // 0.000001 ether, 1e12 wei
+    uint256 public constant CYCLE_DURATION = 20 days;
+    bytes32 constant campaignManagerRole = keccak256("CAMPAIGN_MANAGER_ROLE");
+    UD60x18 rateOfDecay = ud(0.99e18);
+
+    // DAC Variables
+    uint256 public projectID = 5;
+    uint256 public cycleExpiryDate;
+
+    // Events
+    event RefundBonusDeposited(address _campaignManager, uint256 indexed _totalRefundBonus);
+    event PledgeMade(address, uint256);
+    event CycleHasClosed(bool indexed, bool indexed);
+    event CycleRefundBonusWithdrawal(address indexed, uint256 indexed);
+    event CreatorWithdrawal(address, uint256);
+
+    function setUp() external {
+        vm.deal(campaignManager, STARTING_BALANCE);
+        vm.deal(admin, STARTING_BALANCE);
+        vm.deal(pledger1, STARTING_BALANCE);
+        vm.deal(pledger2, STARTING_BALANCE);
+        vm.deal(pledger3, STARTING_BALANCE);
+        vm.deal(rando, STARTING_BALANCE);
+
+        vm.etch(address(controller), "I'm the operator with my pocket calculator");
+        vm.etch(address(directory), "I'm the operator with my pocket calculator");
+        vm.etch(address(fundAccessConstraintsStore), "I'm the operator with my pocket calculator");
+        vm.etch(address(paymentTerminalStore), "I'm the operator with my pocket calculator");
+        vm.etch(address(ethPaymentTerminal), "I'm the operator with my pocket calculator");
+
+        _terminals.push(ethPaymentTerminal);
+
+        // Deploy the target contract.
+        mockExternalCallsForConstructor();
+        vm.prank(admin);
+        dominantJuice =
+        new DominantJuice(campaignManager, projectID, CYCLE_TARGET, MIN_PLEDGE_AMOUNT, controller, paymentTerminalStore);
+
+        // Get cycleExpiryDate for use in tests
+        DominantJuice.Campaign memory campaign = dominantJuice.getCampaignInfo();
+        cycleExpiryDate = campaign.cycleExpiryDate;
+    }
+
+    /**
+     * List of Helper Functions (Helper section is at end of this contract body):
+     * - mockExternalCallsForConstructor(): mocks all external calls needed in the constructor
+     * - pledge(): makes a pledge for the input pledger with the input amount.
+     * - successfulCycleHasExpired(): Mocks 3 pledgers pledging over target amount and advances time past cycleExpiryDate
+     * - failedCycleHasExpired(): Mocks 3 pledgers pledging under target amount and advances time past cycleExpiryDate
+     * - redeem(): requests a redemption for the input redeemer.
+     */
+
+    //////////////////////////////////
+    // supportsInterface() Test
+    //////////////////////////////////
+
+    // PASSING
+    function test_supportsInterface_happyPathForAllThree() public {
+        // Interface IDs from Juicebox docs: https://docs.juicebox.money/dev/build/namespace/
+        bytes4 dataSourceID = 0x71700c69;
```

Comment:
I guess I haven’t really understood the returns of supportsInterface. Someone else told me this too, but then aren’t you just checking if `type(IJBFundingCycleDataSource3_1_1).interfaceId` == `type(IJBFundingCycleDataSource3_1_1).interfaceId`. What does that actually do?

Reply:
you're testing to make sure that the contract actually does what you expect it to do, that's all. You just want to make sure you have remembered each interface here, not what the actual interface ID values are. If I'm the one creating the interface IDs (ie they are my interfaces, not something from a library), I usually have a separate test to verify the interface IDs are actually what I expect them to be so that I don't inadvertently change what `type(IMyInterface).interfaceId` returns.

The reason for this in general is that you may use an interface 10 times, so you only want one test to fail if you change what the value of the interface ID is, not 10 different tests to fail, because it's less clear what is wrong. You want one of the 10 tests to fail if you accidentally remove the interface from one of the contracts later and you shouldn't have. That's the separation of test logic. It seems like overkill because you have one contract here, but if you play this game long enough, you'll have hundreds of contracts and they will reuse these interfaces a lot.

I'm also trying to teach you good practices, and you can decide to ignore the advice for this derivative use case if you want and just take the knowledge with you for later.

Reply:
btw, this really only works to reuse this code because you're in solidity. I have to hard-code them in Typescript tests, and that's just life.

Reply:
Thank you, and I appreciate you teaching me good practices! The logic all makes sense, and I understand the intent, but in the actual `supportsInterface` function it seems to me that what we are doing is the same as asking does `square == square`? Isn't it going to always return true? Coded as it is, I don't understand what it's actually _checking_.

Reply:
you're literally checking that you didn't forget it and that someone doesn't inadvertently remove it when they change the code down the line. Some tests are just there to prevent bone-headed mistakes in the future, and really that's all this test is and should be doing. There's no real logic in the underlying function.